content/blog/2023-12-11-A DSL for Metastability.md

```diff
@@ -0,0 +1,237 @@
+        t_t = 0
+        r = {type: "routing message", payload: "-", destination: BROADCAST};
+        if not outq.is_full {
+            put(outq, r);
+        } else {
+            nop;
+        }
+    } else {
+        nop;
+    }
+}
+
+### Send
+while not outq.is_empty:
+    m = get(outq);
+    give(m);
+```
+First, some notes:
+1. $\texttt{take}$ and $\texttt{give}$ are keywords to interact with the network. I did not write them in the grammar as I am still not sure how to treat them. Informally, $\texttt{take}$ receives the next request to be delivered by the network, and $\texttt{give}$ gives a request from the output queue to the network.
+2. $\texttt{BROADCAST}$ is a keywork to denote that this request must be sent to all of the agents.
+3. $r.f$ is sugar to access the field $f$ of request $r$, and $q.$is_full is sugar for checking the size of the queue (similarly for is_empty).
+
+The example here is from a [paper by Floyd and Van Jacobson](https://ee.lbl.gov/papers/sync_94.pdf).
+The paper studies synchronization of periodic routing messages.
+In a nutshell, this is what happens:
+1. Each agent, upon the expiration of its $T_t$ timer enters a listening state, during which it listens for routing messages from other nodes and processes them. Moreover, it sends them a routing message once it enters the listening state.
+2. For each message that it receives during the listening state, it spends an extra step in this state, captured by the increment to the timer for this state, $T_p$.
+3. Once the $T_p$ timer expires, the agent enters the idle state in which it is waiting for the expiration of the $T_t$ timer, captured by incrementing $t_t$. If it receives messages during this period, it just ignores them.
+4. There is a special message of the type "update", that can be injected anytime by the environment. This message abruptly takes the agent to the listening state, no matter the value of the timers.
+
+Metastability in this example happens when a triggering update message is injected into one of the agents, possibly by a network operator playing the role of the environment.
+As a result, all of the agents are eventually synchronized, meaning that they send their routing messages at axactly the same time.
+This is undesired behavior in a realistic network, as it introduces bursts of packets periodically.
+As an analogy to the retry storm example, the self-sustaining loop here is the synchronous firing of routing messages.
+
+This instance of metastability is particularly robust, as it does not depend on concurrency at all: no matter what the ordering of messages is, the system converges to a bad self-sustaining state after the trigger.
+
+# Implementation
+I decided to implement this DSL in Rust.
+The main reason was that I wanted to learn Rust, and after spending countless hours wrestling with the borrow-checker, I can confirm that one should not toy around with Rust.
+I am, by the way, still wrestling as we speak.
```

Comment:
It can be a serious challenge! But people often find that it can be worth the struggle in the end…